### PR TITLE
ci: add functional test using retail-demo data

### PR DESCRIPTION
**Pull request by kalensk**
This adds an automated functional test for loading and upgrading the retail-demo data set. The two main commits go together, but I split them up for easy reviewing.

During development there were a few notes such as:
- Why do we need to restart in order to have the bm.so function take affect? [Postgres C-Language Functions Documentation](https://www.postgresql.org/docs/9.4/xfunc-c.html)
- There are a few workarounds for the info_rels error we were seeing. We were not able to reliably reproduce this issue outside of the CI testing development cycle. The error occurred during pg_upgrade check as shown:
> SQL command failed
CREATE TEMPORARY TABLE info_rels (reloid, indtable, toastheap) AS SELECT c.oid, i.indrelid, 0::oid FROM pg_catalog.pg_class c JOIN pg_catalog.pg_namespace n       ON c.relnamespace = n.oid LEFT OUTER JOIN pg_catalog.pg_index i         ON c.oid = i.indexrelid WHERE relkind IN ('r', 'o', 'b', 'i', 'm') AND   i.indisvalid IS DISTINCT FROM false AND i.indisready IS DISTINCT FROM false AND      ((n.nspname !~ '^pg_temp_' AND     n.nspname !~ '^pg_toast_temp_' AND     n.nspname NOT IN ('pg_catalog', 'information_schema',
'binary_upgrade', 'pg_toast') AND     n.nspname NOT IN ('gp_toolkit', 'pg_bitmapindex', 'pg_aoseg') AND           c.oid >= 16384)   OR (n.nspname = 'pg_catalog' AND     relname IN ('pg_largeobject', 'pg_largeobject_loid_pn_index') ));
ERROR:  could not create relation file 'base/16384/17214', relation name 'info_rels': File exists

The workaround for the above info_rels issue is to perform a restart after loading the retail demo-data, and then a gpcheckcat after applying the upgrade fixups. See comments in code.

[Test Pipeline](https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/gpupgrade:retail-demo)

### CODE REVIEW

File: ci/generated/prod-pipeline.yml [hunk 1]
```diff
@@ -656,6 +770,7 @@ jobs:
         - 5-to-6-link-mode
         - 5-to-6-primaries-only
         - 5-to-6-no-standby
+        - 5-to-6
```
Comment (marksliva) [inline ci/generated/prod-pipeline.yml:773]: Is there a reason `5-to-6-retail-demo` is not in this list of passed constraints for publishing the release candidate?

File: ci/scripts/filter/filter.go [hunk 2]
```diff
@@ -29,8 +29,18 @@ import (
...
+	Replacement string
+}
+
+func (t *Replacer) replace(line string) string {
+	return t.Regex.ReplaceAllString(line, t.Replacement)
+}
+
 var lineRegexes []*regexp.Regexp
 var blockRegexes []*regexp.Regexp
+var replacments []*Replacer
```
Comment (marksliva) [inline ci/scripts/filter/filter.go:43]: There's a typo in the variable name.

File: ci/scripts/filter/filter.go [hunk 3]
```diff
@@ -46,12 +56,22 @@ func init() {
 		"COMMENT ON DATABASE postgres IS",
 	}
 
+	// replacementPatterns is a map of regex substitutions.
+	replacementPatterns := []map[string]string{
```
Comment (marksliva) [inline ci/scripts/filter/filter.go:60]: Why is this a list of maps? A simple map of string -> string would suffice, and then we can get rid of the double for loop.
```
replacementPatterns := map[string]string{
    {"WITH \\(tablename='(.+)', appendonly='true', compresstype=(.+), orientation='column' \\)": "WITH (tablename='${1}', appendonly=true, compresstype=${2}, orientation=column )"},
}
...
for regex, replacement := range replacementPatterns {
...
```

File: ci/generated/dev-pipeline.yml [hunk 4]
```diff
@@ -631,3 +638,110 @@ jobs:
...
+        - get: gpupgrade_src
+          trigger: true
+        - get: bin_gpdb_old
+          resource: bin_gpdb5_centos6
+          trigger: true
+        - get: bin_gpdb_new
+          resource: bin_gpdb6_centos6
+          trigger: true
+        - get: ccp_src
+        # FIXME: this is not guaranteed to be the same dump generated by the
```
[the remark this answers is not in the document]
Comment (kalensk) [reply ci/generated/dev-pipeline.yml]: Fixed. My if statement in the template.yml should have included the comment. Thanks.

File: ci/generated/prod-pipeline.yml (hunk 1 above)
Comment (kalensk) [reply ci/generated/prod-pipeline.yml]: Fixed. 

I didn't realize this was a thing (especially this was just recently added). Thanks for the reminder.

File: ci/scripts/filter/filter.go (hunk 2 above)
Comment (kalensk) [reply ci/scripts/filter/filter.go]: fixed

File: ci/scripts/filter/filter.go (hunk 3 above)
Comment (kalensk) [reply ci/scripts/filter/filter.go]: Fixed.

**Comment by bhuvnesh2703** [comment]
LGTM, but it requires some conflict resolution..

File: ci/template.yml [hunk 5]
```diff
@@ -248,6 +257,10 @@ jobs:
           {{- else }}
           resource: dump_gpdb6_icw_gporca_centos6
           {{- end }}
+        {{- end }}
+        {{- if .RetailDemo }}
+        - get: retail_demo_src
+        {{- end }}
```
Comment (jchampio) [inline ci/template.yml]: nit: `{{ if .RetailDemo }}` / `{{ else }}` would make the intent clearer here, I think.

File: ci/template.yml [hunk 6]
```diff
@@ -305,8 +318,30 @@ jobs:
           args:
             - greenplum-db-{{majorVersion .Source}}
             - greenplum-db-{{majorVersion .Target}}
+    {{- if not .RetailDemo }}
     - task: load_dump
       file: gpupgrade_src/ci/tasks/load-dump.yml
+    {{- end }}
+    {{- if .RetailDemo }}
```
Comment (jchampio) [inline ci/template.yml]: Ditto.

File: ci/template.yml [hunk 7]
```diff
@@ -305,8 +318,30 @@ jobs:
...
+        inputs:
+          - name: gpupgrade_src
+          - name: retail_demo_src
+          - name: ccp_src
+          - name: cluster_env_files
+        run:
+          path: gpupgrade_src/ci/scripts/load-retail-data.bash
+      params:
+        OLD_PACKAGE: greenplum-db-5
+        NEW_PACKAGE: greenplum-db-6
```
Comment (jchampio) [inline ci/template.yml:345]: Why are `OLD_PACKAGE` and `NEW_PACKAGE` resurrected here?

edit: oh, see below.

File: ci/scripts/load-retail-data.bash [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+# Retrieves the installed GPHOME for a given GPDB RPM.
+rpm_gphome() {
+    local package_name=$1
+
+    local version=$(ssh -n mdw rpm -q --qf '%{version}' "$package_name")
+    echo /usr/local/greenplum-db-$version
+}
+
+export GPHOME_OLD=$(rpm_gphome ${OLD_PACKAGE})
+export GPHOME_NEW=$(rpm_gphome ${NEW_PACKAGE})
```
Comment (jchampio) [inline ci/scripts/load-retail-data.bash:19]: This logic is no longer necessary. Use `/usr/local/greenplum-db-old` and `/usr/local/greenplum-db-new` as GPHOMEs.

File: ci/scripts/filter/filter.go [hunk 9]
```diff
@@ -46,12 +56,20 @@ func init() {
 		"COMMENT ON DATABASE postgres IS",
 	}
 
+	// replacementPatterns is a map of regex substitutions.
+	replacementPatterns := map[string]string{
+		"WITH \\(tablename='(.+)', appendonly='true', compresstype=(.+), orientation='column' \\)": "WITH (tablename='${1}', appendonly=true, compresstype=${2}, orientation=column )",
```
Comment (jchampio) [inline ci/scripts/filter/filter.go:61]: nit: backtick-quoted strings are useful to reduce the escaping.

File: ci/scripts/filter/filter.go [hunk 10]
```diff
@@ -29,8 +29,18 @@ import (
 	"strings"
 )
 
+type Replacer struct {
+	Regex       *regexp.Regexp
+	Replacement string
+}
+
+func (t *Replacer) replace(line string) string {
```
Comment (jchampio) [inline ci/scripts/filter/filter.go:37]: nit: an exported type with a single unexported method is a little strange. Is the linter happy with this?

File: ci/scripts/filter/filter.go [hunk 11]
```diff
@@ -46,12 +56,20 @@ func init() {
...
+	}
+
 	for _, pattern := range linePatterns {
 		lineRegexes = append(lineRegexes, regexp.MustCompile(pattern))
 	}
 	for _, pattern := range blockPatterns {
 		blockRegexes = append(blockRegexes, regexp.MustCompile(pattern))
 	}
+	for regex, replacement := range replacementPatterns {
+		replacements = append(replacements, &Replacer{Regex: regexp.MustCompile(regex), Replacement: replacement})
```
Comment (jchampio) [inline ci/scripts/filter/filter.go:71]: nit: this line is getting long.

File: ci/scripts/filter/filter.go [hunk 12]
```diff
@@ -94,6 +112,10 @@ nextline:
 			}
 		}
 
+		for _, r := range replacements {
+			line = r.replace(line)
+		}
```
Comment (jchampio) [inline ci/scripts/filter/filter.go:117]: How does this perform in practice, both for the retail-demo data and 6X ICW? The new replacement regex doesn't have any anchors to help the regex engine fail fast.

File: ci/template.yml (hunk 5 above)
Comment (kalensk) [reply ci/template.yml]: Updated during the rebase.

File: ci/template.yml (hunk 6 above)
Comment (kalensk) [reply ci/template.yml]: Updated during the rebase.

File: ci/scripts/filter/filter.go (hunk 10 above)
Comment (jchampio) [reply ci/scripts/filter/filter.go:37]: Never mind, it's the reverse (exported method on an unexported type) that causes issues.

File: ci/scripts/filter/filter.go (hunk 12 above)
Comment (kalensk) [reply ci/scripts/filter/filter.go:120]: If I remember correctly, it was about ~17 seconds without anchoring, and ~13 seconds with.

File: ci/scripts/filter/filter.go (hunk 11 above)
Comment (kalensk) [reply ci/scripts/filter/filter.go]: fixed.

File: ci/scripts/filter/filter.go (hunk 9 above)
Comment (kalensk) [reply ci/scripts/filter/filter.go]: udpated.

File: ci/scripts/load-retail-data.bash (hunk 8 above)
Comment (kalensk) [reply ci/scripts/load-retail-data.bash]: Updated.

**Pull request merged**